### ISSUE TITLE
Creating custom loss function

### ISSUE THREAD

**Issue by devinbostIL** [OpenNMT/OpenNMT-py#108]
I was given some direction here: [Creating custom cost metric](http://forum.opennmt.net/t/creating-custom-cost-metric/603/2) for how to create a custom Criterion in the Torch version of OpenNMT so that I can implement a custom loss function (other than perplexity) because our data science team realized that perplexity alone wasn't capturing some essential information during training for our particular use cases.
However, while going through the PyOpenNMT code base, I'm noticing some differences regarding how loss/cost is calculated during training. 

In the Loss file here: [PyOpenNMT Loss file with related classes](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Loss.py)
I noticed that there are several classes, and without spending a lot of time trying to disambiguate (and risk that they could all end up getting changed as the PyOpenNMT architecture is quickly evolving), I'd like some guidance on what I would need to define/override in order to implement custom loss functions. 
I was initially perhaps naively thinking that I could simply create a new class that implements the loss math calculation in a forward method (using matrix math) and allow the awesome PyTorch autograd features to handle the backprop calculation, like here: [Creating custom loss function in PyTorch](https://discuss.pytorch.org/t/build-your-own-loss-function-in-pytorch/235/38), perhaps like the implementation of Linear here: [PyTorch linear module](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/linear.py).
But after I was told by the OpeNMT folks that I would need to create a Criterion implementation, I came across this: [PyTorch ClassNLLCriterion](https://github.com/pytorch/pytorch/blob/master/torch/lib/THNN/generic/ClassNLLCriterion.c)
which appears to be (and please excuse my ignorance) largely dependent on the C CUDA functions that I'm not particularly familiar with. 

So, could someone please give me some guidance (and I apologize if this is not the correct place for me to bring this subject up) on how I might implement a custom loss function in PyOpenNMT? I am willing to fork and submit a PR if I need to help with refactoring to make it possible, but I'd just like some direction.

**Comment by falcondai** [comment]
Due to the alpha developmental stage we are in, as you are aware, I would suggest you to fork the repo and change the loss function to your liking. These are some pointers:

you might want to directly modify the criterion calculation (note line 17) inside this function
https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Loss.py#L11

the definition is invoked in `train.py`
https://github.com/OpenNMT/OpenNMT-py/blob/master/train.py#L211

> which appears to be (and please excuse my ignorance) largely dependent on the C CUDA functions that I'm not particularly familiar with.

unless you have strict demand for efficiency, you can probably find a way to implement your loss in python using existing functions in pyTorch.

**Comment by srush** [comment]
Hi @devinbostIL, 

We can help with this. Can you specify mathematically the loss function that you would like? We can give directions on how to add it.

**Comment by devinbostIL** [comment]
Ideally, we would like to implement a combined metric that relates the Google-BLEU score (called the "GLEU" score) with a word accuracy score (calculated as WAcc = 1 - WER, where WER is the Word Error Rate) in a manner like an F-score:

2 * [ (GLEU * WAcc) / (GLEU + WAcc) ].

**Comment by srush** [comment]
Oh, so you can't easily train on that, so I wouldn't call it a loss function. 

Since we don't actually do translation and produce full output in the training loop, this is unlikely to be added. I would recommend using translate to produce text outputs and then just using that as a metric.

**Comment by devinbostIL** [comment]
Is the issue that the proposed score is calculated on a phrase-basis whereas we need to train on something that works on a word-basis? I'd like to better understand what the issue is.

**Comment by devinbostIL** [comment]
How about implementing Kullback–Leibler divergence as a cost function?

**Comment by srush** [comment]
- Yes, it needs to be word based. There is a lot written on this, but it comes down to word-based methods not requiring sampling. 
- Log-likelihood is a version of KL. (KL is non-symmetric so you can flip it, but it becomes unstable the other way.)

**Comment by devinbostIL** [comment]
Thanks for the feedback. Do you have any references handy so that I can research this further (regarding word-based methods not requiring sampling)?

**Comment by srush** [comment]
Going to close this for now. It should be pretty straightforward to add a new loss function in `Loss.py` now
